src/wasm-emscripten.cpp

```diff
@@ -196,6 +196,96 @@ void EmscriptenGlueGenerator::generateDynCallThunks() {
   }
 }
 
+struct JSCallWalker : public PostWalker<JSCallWalker> {
+  Module &wasm;
+  JSCallWalker(Module &_wasm) : wasm(_wasm) {
+    if (wasm.table.segments.size() == 0) {
+      auto emptySegment =
+          wasm.allocator.alloc<Const>()->set(Literal(uint32_t(0)));
+      wasm.table.segments.emplace_back(emptySegment);
+    }
+    std::vector<Name> tableSegmentData = wasm.table.segments[0].data;
+
+    // Check if jsCalls have already been created
+    for (unsigned i = 0; i < tableSegmentData.size(); ++i) {
+      if (tableSegmentData[i].startsWith("jsCall_")) {
+        jsCallStartIndex = i;
+        return;
+      }
+    }
+    jsCallStartIndex = wasm.table.segments[0].data.size();
+  }
+  void visitCallIndirect(CallIndirect *curr);
+
+  bool createJSCallThunks;
+  unsigned jsCallStartIndex;
+  // Function type signatures used in call_indirect instructions
+  std::set<std::string> indirectlyCallableSigs;
+};
+
+// Gather all function signatures used in call_indirect, because any of them
+// can be used to call function pointers created by emscripten's addFunction.
+void JSCallWalker::visitCallIndirect(CallIndirect* curr) {
+  indirectlyCallableSigs.insert(getSig(wasm.getFunctionType(curr->fullType)));
```

Comment:
This is short enough that it might as well be inline in JSCallWalker?

Reply:
Done